### ISSUE TITLE
Include system metadata in user-agent

### ISSUE THREAD

**Issue by alex** [pypa/setuptools#1066]
pip does this in order to provide useful metadata for the server on the platform that's being installed from: https://github.com/pypa/pip/blob/master/pip/download.py#L64-L130

It'd be great if `setuptools` did the same thing.

(Until `setuptools` downloading functionality is totally removed in the bright and glorious future of course :-))

cc: @dstufft

**Comment by dstufft** [comment]
If setuptools wants to do this, and generally adopted the same format (at least the JSON part), it'd be really easy for PyPI to start reading data from it.

**Comment by alex** [comment]
Updated link: https://github.com/pypa/pip/blob/master/src/pip/_internal/download.py#L65-L131

**Comment by pganssle** [comment]
@alex This is for the download functionality, not upload, right?

**Comment by pganssle** [comment]
Whether it's for download or upload, as far as I know it's functionality that would ideally be removed from setuptools entirely, so I think the big question before we give this the green light is whether it's easier to remove the download functionality entirely or to add this enhancement. Additionally, are there end users who are affected by this and *cannot* switch to `pip` (or `twine`) for some reason?

**Comment by alex** [comment]
Yes, download is the only part I care about. Upload doesn't hurt of course, but I don't care :-)

And yeah, I'd _love_ to just delete the `easy_install` functionality out of setuptools, but right now it's still fairly relied upon for `setup_requires`.

**Comment by benoit-pierre** [comment]
Why not use PEP 518?

**Comment by alex** [comment]
So, speaking for pyca/cryptography (the primary package I care about), our next release will include optional support for PEP 518 -- that is, we'll use both `pyproject.toml` and `setup.py`. We're not yet prepared to break all the old pips (I want to, but @reaperhulk won't let me), so I'd like to have data on what's going on there.

**Comment by pganssle** [comment]
Ok, I'm going to mark this as "Needs Implementation" and say that we will accept a PR that adds this information at least to `easy_install` user agent. I don't think we're adding a *huge* maintenance burden here.

That said, if you are a bright-eyed and bushy-tailed new OSS contributor and want to take on this issue: 1. thank you for your help! 2. this is a fix to functionality that we would like to actively remove, so please do not be offended if you fix this and then the whole section of the code you've fixed is ripped out 6 months later. My only consolation is that given how slowly everything moves in highly depended-upon libraries like `setuptools`, 6 months is probably **wildly** optimistic.

**Comment by dstufft** [comment]
It'll likely require changes to Linehaul as well to parse the new user agent.

We should probably take pip's schema and make a standard of it.
